x-pack/legacy/plugins/file_upload/public/util/file_parser.test.js

```diff
@@ -0,0 +1,133 @@
+
+describe('parse file', () => {
+  const cleanAndValidate = jest.fn(a => a);
+  const resolve = jest.fn();
+  const reject = jest.fn();
+  const previewFunction = jest.fn();
+
+  beforeEach(() => {
+    cleanAndValidate.mockClear();
+    resolve.mockClear();
+    reject.mockClear();
+    previewFunction.mockClear();
+  });
+
+  it('should parse valid JSON', () => {
+    const validJsonFileResult = {
+      target: {
+        result: JSON.stringify({
+          'type': 'Feature',
+          'geometry': {
+            'type': 'Polygon',
+            'coordinates': [[
+              [-104.05, 78.99],
+              [-87.22, 78.98],
+              [-86.58, 75.94],
+              [-104.03, 75.94],
+              [-104.05, 78.99]
+            ]]
+          },
+        })
+      }
+    };
+
+    parseAndClean(cleanAndValidate, resolve, reject)(validJsonFileResult);
+    // Confirm cleanAndValidate called
+    expect(cleanAndValidate.mock.calls.length).toEqual(1);
+    // Confirm resolve called
+    expect(resolve.mock.calls.length).toEqual(1);
+  });
+
+  it('should reject invalid JSON', () => {
+    const inValidJsonFileResult = {
+      target: {
+        result: 'this should not work'
+      }
+    };
+
+    parseAndClean(cleanAndValidate, resolve, reject)(inValidJsonFileResult);
+    // Confirm cleanAndValidate not called
+    expect(cleanAndValidate.mock.calls.length).toEqual(0);
+    // Confirm resolve not called
+    expect(resolve.mock.calls.length).toEqual(0);
+    // Confirm reject called
+    expect(reject.mock.calls.length).toEqual(1);
+  });
+
+  it('should reject object instead of JSON', () => {
+    const objectInsteadOfJson = {
+      target: {
+        result: {
+          'type': 'Feature',
+          'geometry': {
+            'type': 'Polygon',
+            'coordinates': [[
+              [-104.05, 78.99],
+              [-87.22, 78.98],
+              [-86.58, 75.94],
+              [-104.03, 75.94],
+              [-104.05, 78.99]
+            ]]
+          },
+        }
+      }
+    };
+
+    parseAndClean(cleanAndValidate, resolve, reject)(objectInsteadOfJson);
+    // Confirm cleanAndValidate not called
+    expect(cleanAndValidate.mock.calls.length).toEqual(0);
+    // Confirm resolve not called
+    expect(resolve.mock.calls.length).toEqual(0);
+    // Confirm reject called
+    expect(reject.mock.calls.length).toEqual(1);
+  });
+
+  it('should call preview callback function if provided', () => {
+    const justFinalJson = {
+      'type': 'Feature',
+      'geometry': {
+        'type': 'Polygon',
+        'coordinates': [[
+          [-104.05, 78.99],
+          [-87.22,  78.98],
+          [-86.58,  75.94],
+          [-104.03, 75.94],
+          [-104.05, 78.99]
+        ]]
+      },
+    };
+
+    jsonPreview(justFinalJson, previewFunction);
+    // Confirm preview function called
+    expect(previewFunction.mock.calls.length).toEqual(1);
+  });
+
+  it('should use object clone for preview function', () => {
```

Comment:
why? 

Shouldn't it be the opposite? Cloning can be a huge performance hit, and it was the reason for known performance issues in the Maps app already (e.g. https://github.com/elastic/kibana/issues/35618).

Reply:
We're modifying the object downstream, it was causing the error (second checkbox) described in this #38809. " Prevent __kbn__feature_id__ from getting inserted in docs". This is happening downstream [here](https://github.com/elastic/kibana/blob/45e14cf0623077aeefe4c902302309bccc307988/x-pack/legacy/plugins/maps/public/shared/layers/vector_layer.js#L505), updating the object prior to indexing.